src/renderers/dom/server/__tests__/ReactServerRendering-test.js

```diff
@@ -254,6 +254,71 @@ describe('ReactServerRendering', function() {
         'renderToString(): You must pass a valid ReactElement.'
       );
     });
+
+    it('should warn and ignore setState calls executed ' +
+        'after componentWillMount', function() {
+      spyOn(console, 'error');
+
+      var Component = React.createClass({
+        componentWillMount: function() {
+          setTimeout(() => {
+            this.setState({text: 'hello, world'});
+          });
+        },
+        render: function() {
+          return <div></div>;
+        },
+      });
+
+      ReactReconcileTransaction.prototype.perform = function() {
+        // We shouldn't ever be calling this on the server
+        throw new Error('Browser reconcile transaction should not be used');
+      };
+      ReactServerRendering.renderToString(
+        <Component />
+      );
+
+      jest.runAllTimers();
+      expect(setTimeout.mock.calls.length).toBe(1);
+      expect(console.error.calls.length).toBe(1);
+      expect(console.error.argsForCall[0][0]).toContain(
+        'setState(...): method calls executed after componentWillMount ' +
+        '(e.g. setTimeout() callbacks)  are ignored if component was ' +
+        'rendered on server.'
+      );
+    });
+
+    it('should warn and ignore all forceUpdate calls', function() {
+      spyOn(console, 'error');
+
+      var Component = React.createClass({
+        componentWillMount: function() {
+          this.forceUpdate();
+          setTimeout(() => {
+            this.forceUpdate();
+          });
+        },
+        render: function() {
+          return <div></div>;
+        },
+      });
+
+      ReactReconcileTransaction.prototype.perform = function() {
+        // We shouldn't ever be calling this on the server
+        throw new Error('Browser reconcile transaction should not be used');
+      };
+      ReactServerRendering.renderToString(
+        <Component />
+      );
+
+      var expectedError = 'forceUpdate(...): all method calls are ignored ' +
+        'if component was rendered on server.';
+      jest.runAllTimers();
+      expect(setTimeout.mock.calls.length).toBe(1);
+      expect(console.error.calls.length).toBe(2);
```

Comment:
@jimfb I saw this pattern a lot of times, but it seems like it would be much easier to rework an abstract component as a whole if we know about all warnings at once (which can be delayed in time), instead of discovering them one by one. IMHO

Anyways, let me know if we still want to stick to the first warning only policy. I'll just add one more property to the `_serverSideRendered` object in that case